h1b/src/hil/flash/h1b_hw.rs

```diff
@@ -12,256 +12,274 @@
+        /// The last cycle BULKERASE should be asserted during bulkerase.
+        bulkerase_bulkerase_last_cycle: VolatileCell<u32>,
 
-	/// The first cycle MAS1 should be asserted during bulkerase.
-	bulkerase_mas1_first_cycle: VolatileCell<u32>,
+        /// The first cycle MAS1 should be asserted during bulkerase.
+        bulkerase_mas1_first_cycle: VolatileCell<u32>,
 
-	/// The last cycle MAS1 should be asserted during bulkerase.
-	bulkerase_mas1_last_cycle: VolatileCell<u32>,
+        /// The last cycle MAS1 should be asserted during bulkerase.
+        bulkerase_mas1_last_cycle: VolatileCell<u32>,
 
-	/// The first cycle NVSTR should be asserted during bulkerase.
-	bulkerase_nvstr_first_cycle: VolatileCell<u32>,
+        /// The first cycle NVSTR should be asserted during bulkerase.
+        bulkerase_nvstr_first_cycle: VolatileCell<u32>,
 
-	/// The last cycle NVSTR should be asserted during bulkerase.
-	bulkerase_nvstr_last_cycle: VolatileCell<u32>,
+        /// The last cycle NVSTR should be asserted during bulkerase.
+        bulkerase_nvstr_last_cycle: VolatileCell<u32>,
 
-	/// Controller debug signals.
-	debug_signals: VolatileCell<u32>,
+        /// Controller debug signals.
+        debug_signals: VolatileCell<u32>,
 
-	// The integration test controls are omitted because they are unnecessary
-	// and are after a large address space gap.
+        // The integration test controls are omitted because they are unnecessary
+        // and are after a large address space gap.
 }
 
 impl super::hardware::Hardware for H1bHw {
-	fn is_programming(&self) -> bool {
-		// TODO(jrvanwhy): Only checks the second flash bank.
-		self.pe_control_1.get() != 0
-	}
-
-	fn read(&self, offset: usize) -> u32 {
-		// The two flash macros are in consecutive memory locations, so they can
-		// be addressed as one.
-		unsafe { ::core::ptr::read_volatile((0x40000 as *const u32).add(offset)) }
-	}
-
-	fn read_error(&self) -> u16 {
-		self.error_code.get() as u16
-	}
-
-	fn set_transaction(&self, offset: usize, size: usize) {
-		use self::TransactionParameters::{Offset,Size};
-		// The offset is relative to the beginning of the flash module. There
-		// are 128 pages per flash module.
-		// TODO(jrvanwhy): Assumes the read is from the second flash bank.
-		let offset = offset - 128 * super::WORDS_PER_PAGE;
-		self.transaction_parameters.write(Offset.val(offset as u32) + Size.val(size as u32));
-	}
-
-	fn set_write_data(&self, data: &[u32]) {
-		for (i, &v) in data.iter().enumerate() { self.write_data[i].set(v); }
-	}
-
-	fn trigger(&self, opcode: u32) {
-		self.program_erase_enable.set(0xb11924e1);
-		// TODO(jrvanwhy): Assumes the write is to the second flash bank (where
-		// the nvmem counter is).
-		self.pe_control_1.set(opcode);
-	}
+        fn is_programming(&self) -> bool {
+                // TODO(jrvanwhy): Only checks the second flash bank.
+                self.pe_control_1.get() != 0
+        }
+
+        fn read(&self, offset: usize) -> u32 {
```

Comment:
Read should probably return a ReturnCode instead of silently reading 0.

Reply:
I don't think the overhead is significant, and should generally optimize away anyway. Another option is to have a read_with_default method instead, so callers pick the value rather than it being hardcoded to zero.

If you're still concerned about the overhead, though, let me know and I'll approve as-is.

Reply:
It now returns a ReturnCode, with ReturnCode::ESIZE on an out-of-bounds offset.